# 1 SSRå¼•ç‰©æ‰¹é‡è®¾è®¡
##### ğŸ™ˆ: yukaiquan
##### ğŸ“§: 1962568272@qq.com
## 1.1 é€šè¿‡misaç»“æœåˆ¶ä½œbedæ–‡ä»¶ä»¥åŠå‰”é™¤ä¸€äº›è¿ç»­å¤šæ®µçš„SSR
é‚£ä¹ˆå†™ä¸€ä¸ªpythonè„šæœ¬å§
```python
# yukaiquan
# email:1962568272@qq.com
import pandas as pd
import sys


input_file = sys.argv[1]
output_file = sys.argv[2]
output_bed = sys.argv[3]
misa_df = pd.read_csv(input_file, sep='\t')
misa_df = misa_df[misa_df['SSR type'].isin(['p2', 'p3', 'p4', 'p5', 'p6'])]
# output
misa_df.to_csv(output_file, sep='\t', index=False)

misa_df = misa_df[['ID', 'SSR nr.']]

misa_df['name'] = misa_df['ID'].str.replace(
    ':', '_').str.replace('-', '_') + '_' + misa_df['SSR nr.'].astype(str)

misa_df['chr'] = misa_df['ID'].str.split(':').str.get(0)
misa_df['start'] = misa_df['ID'].str.split(
    ':').str.get(1).str.split('-').str.get(0)
misa_df['end'] = misa_df['ID'].str.split(
    ':').str.get(1).str.split('-').str.get(1)

misa_df = misa_df[['chr', 'start', 'end', 'name']]
# output
misa_df.to_csv(output_bed,
               sep='\t', index=False, header=False)

```
è¿è¡Œä¸€ä¸‹`python misa_result_2_bed.py Avena_eriantha.fasta.gz_second.tsv Avena_eriantha.fasta.gz_second_vp1c.misa Avena_eriantha.fasta.gz_second_vp1c.bed`
## 1.2 æ ¹æ®bedæ–‡ä»¶åˆ¶ä½œfastaæ–‡ä»¶
```shell
bedtools getfasta -fi ACD_sativa_sfs_genome.fasta -bed /mnt/e/oatdatabase/05_SSR/01_SFS/SFSchr.fasta_second_vp1c.fasta.bed -nameOnly -fo SFSchr.fasta_second_vp1c.fasta
```
## 1.2 åˆ©ç”¨fastaè¿›è¡Œå¼•ç‰©è®¾è®¡
å› ä¸ºæ•°æ®éå¸¸åºå¤§ï¼Œå¦‚æœåˆ©ç”¨primer3å•è¿›ç¨‹è®¾è®¡å¼•ç‰©è‚¯å®šæ˜¯ä¸å¯è¡Œçš„ï¼Œæˆ‘ä»¬éœ€è¦è¿›è¡Œæ”¹é€ è®©primer3æ”¯æŒå¤šè¿›ç¨‹ã€‚äºæ˜¯å†™äº†ä¸€ä¸ªå¤šè¿›ç¨‹çš„pyç¨‹åºã€‚
```python
# yukaiquan
# email:1962568272@qq.com


import primer3
import pandas as pd
import sys
import argparse
import os
import time
from progressbar import ProgressBar
from multiprocessing import Pool
from functools import partial

# primer_condition


def main():
    start = time.time()
    args = check_options(get_options())
    print(args.input + ' ' + str(args.num) + ' ' + str(args.min_size) + ' ' + str(args.max_size) + ' ' +
          str(args.min_tm) + ' ' + str(args.max_tm) + ' ' + str(args.min_gc) + ' ' + str(args.max_gc) + ' ' + str(args.thread) + ' ' + args.output)
    global_args = gloabal_args(int(args.num), int(args.min_size), int(args.max_size), int(
        args.min_gc), int(args.max_gc), int(args.min_tm), int(args.max_tm))
    # primer3_py è¿›è¡Œå¼•ç‰©è®¾è®¡ï¼Œè¾“å‡ºç»“æœæ–‡ä»¶####primer3æ¨¡å—
    # read fasta
    pbar = ProgressBar()
    pbar.start()
    with open(args.input, 'r') as f:
        lines = f.readlines()
        index, seq = readfasta(lines)
    num_treads = int(args.thread)
    # åˆ†å‰²æ•°æ®
    n = len(index) // num_treads
    print(len(index))
    indexs = [index[int(i):int(i + n)] for i in range(0, len(index), n)]
    seqs = [seq[int(i):int(i + n)] for i in range(0, len(seq), n)]
    pool = Pool(processes=num_treads)
    primer_result = pd.DataFrame()
    partial_func = partial(
        primer3_design, global_args=global_args)
    # primer_dfs = pd.DataFrame()
    for i, j in zip(indexs, seqs):
        zip_args = list(zip(i, j))
        primer_df = pool.starmap_async(
            partial_func, zip_args)
        # primer_df = pd.concat([primer_dfs, primer_df])
        # print(primer_df)
        # print(primer_df.get())
        for k in primer_df.get():
            primer_result = pd.concat([primer_result, k])
        # primer_result = primer_result.append(k)
    pool.close()
    pool.join()
    # primer_result = primer3_design(index, seq, global_args)
    primer_result.to_csv(args.output, sep="\t")
    end = time.time()
    print("time: " + str(end - start))
    pbar.finish()
    print("finish")


def primer3_design(index, seq, global_args):
    # build table
    index = index.split("\n")[0]
    seq = seq.split("\n")
    # primer_df = pd.DataFrame()
    # primer finder, dic -> datafrme
    # for i in range(len(index)):
    seq_args = {
        'SEQUENCE_ID': str(index),
        'SEQUENCE_TEMPLATE': str(seq),
        'SEQUENCE_INCLUDED_REGION': [0, len(str(seq))-1],
    }
    GeneID = str(index)

    primer3_result = primer3.bindings.designPrimers(seq_args, global_args)
    # print(primer3_result)
    # change dic
    primer3_result_table_dict = {}
    for j in range(primer3_result["PRIMER_PAIR_NUM_RETURNED"]):
        primer_id = str(j)
        for key in primer3_result:
            if primer_id in key:
                # è¦å°†æ¯ä¸ªä¿¡æ¯ä¸­çš„æ•°å­—å’Œä¸‹åˆ’çº¿å»æ‰
                info_tag = key.replace("_" + primer_id, "")
                # å°±æ˜¯æŠŠä¸åŒçš„å¼•ç‰©å¯¹ç»“æœå½’åˆ°ä¸€èµ·
                try:
                    primer3_result_table_dict[info_tag]
                except:
                    primer3_result_table_dict[info_tag] = []
                finally:
                    primer3_result_table_dict[info_tag].append(
                        primer3_result[key])

    df_index = []

    # append dataframe
    for m in range(primer3_result["PRIMER_PAIR_NUM_RETURNED"]):
        df_index.append(GeneID + "_" + str(m + 1))
    primer3_result_df = pd.DataFrame(
        primer3_result_table_dict, index=df_index)
    # primer_df = pd.concat([primer_df, primer3_result_df])
    # primer_df = primer_df.append(primer3_result_df)
    return primer3_result_df


def readfasta(lines):
    seq = []
    index = []
    seqplast = ""
    numlines = 0
    for i in lines:
        if ">" in i:
            index.append(i.replace("\n", "").replace(">", ""))
            seq.append(seqplast.replace("\n", ""))
            seqplast = ""
            numlines += 1
        else:
            seqplast = seqplast + i.replace("\n", "")
            numlines += 1
        if numlines == len(lines):
            seq.append(seqplast.replace("\n", ""))
    seq = seq[1:]
    return index, seq


def gloabal_args(return_num, min_size, max_size, min_gc, max_gc, min_tm, max_tm):
    global_args = {
        'PRIMER_NUM_RETURN': return_num,  # number of primer return
        ### å¼•ç‰© TM å€¼è®¾å®š: ###
        # TM çš„è®¡ç®—æ–¹æ³•ï¼Œ 1 è¡¨ç¤ºä½¿ç”¨ the SantaLucia parameters (Proc Natl Acad Sci 95:1460-65)
        'PRIMER_TM_FORMULA': 1,
        'PRIMER_MIN_TM': min_tm,  # 55
        'PRIMER_OPT_TM': int((max_tm + min_tm)/2),
        'PRIMER_MAX_TM': max_tm,  # 65
        'PRIMER_PAIR_MAX_DIFF_TM': 2.0,         # ä¸¤ä¸ªå¼•ç‰©ä¹‹é—´çš„ TM å€¼æœ€å¤šç›¸å·® 5 æ‘„æ°åº¦
        'PRIMER_WT_TM_LT': 0,
        'PRIMER_WT_TM_GT': 0,
        'PRIMER_PAIR_WT_DIFF_TM': 0.0,
        'PRIMER_MIN_SIZE': min_size,  # 18
        'PRIMER_OPT_SIZE': int((max_size + min_size)/2),
        'PRIMER_MAX_SIZE': max_size,  # 22
        'PRIMER_WT_SIZE_LT': 0.0,
        'PRIMER_WT_SIZE_GT': 0.0,
        # GCå«é‡æ§åˆ¶
        'PRIMER_MIN_GC': min_gc,  # minimum GC content 40
        'PRIMER_MAX_GC': max_gc,  # maximum GC content 65
        # å¼•ç‰©çƒ­åŠ›å­¦è®¡ç®—#
        # å¼€å¯çƒ­åŠ›å­¦è®¡ç®—ï¼Œå¼€å¯åï¼Œæ ¹æ®çƒ­åŠ›å­¦çš„å€¼ TH å€¼æ¥è®¡ç®—ç½šåˆ†ã€‚ TH çš„ç½šåˆ†æ–¹æ³•ä¸ºï¼šç½šåˆ†ç³»æ•° * (1 / (å¼•ç‰©TM - 4 - THå€¼))ã€‚
        # è¯¥æ–¹æ³•å¥½å¤„æ˜¯ï¼ŒTH å€¼è¶Šå¤§ï¼Œç½šåˆ†åŠ›åº¦è¶Šé‡ã€‚
        'PRIMER_THERMODYNAMIC_OLIGO_ALIGNMENT': 1,  # alignment of oligo
        # å¼•ç‰©è‡ªèº«è¿›è¡Œåå‘äº’è¡¥ï¼Œç½šåˆ†ç³»æ•°è®¡ç®—ä¸º 9 / ( 1/(60-4-45) - 1/(60-4-0) ) = 123.2
        # è¿™æ ·è®¡ç®—ç½šåˆ†ç³»æ•°çš„åŸåˆ™æ˜¯ï¼Œè‹¥ TM ä¸ºæœ€ä½³ 60 æ‘„æ°åº¦çš„æ—¶å€™ï¼Œæ‰€å…è®¸çš„æœ€å¤§ç½šåˆ†å€¼å‡å»æœ€å°ç½šåˆ†å€¼ä¸º 9ï¼Œå’Œ 3' ç«¯ç¢±åŸºçš„ç¨³å®šæ€§çš„ç½šåˆ†é¢åº¦ä¸€è‡´ã€‚
        'PRIMER_MAX_SELF_ANY': 8,
        'PRIMER_WT_SELF_ANY': 0.0,
        'PRIMER_MAX_SELF_ANY_TH': 45.0,
        'PRIMER_WT_SELF_ANY_TH': 123.2,
        # å¼•ç‰©è‡ªèº«è¿›è¡Œ 3' ç«¯åå‘äº’è¡¥å½¢æˆå¼•ç‰©äºŒèšä½“ï¼Œç½šåˆ†ç³»æ•°è®¡ç®—ä¸º 9 / ( 1/(60-4-35) - 1/(60-4-0) ) = 302.4
        'PRIMER_MAX_SELF_END': 3,
        'PRIMER_WT_SELF_END': 0.0,
        'PRIMER_MAX_SELF_END_TH': 35.0,
        'PRIMER_WT_SELF_END_TH': 302.4,
        # left primer å’Œ right primer åºåˆ—çš„åå‘äº’è¡¥
        'PRIMER_PAIR_MAX_COMPL_ANY': 8,
        'PRIMER_WT_PAIR_COMPL_ANY': 0.0,
        'PRIMER_PAIR_MAX_COMPL_ANY_TH': 45.0,
        'PRIMER_WT_PAIR_COMPL_ANY_TH': 123.2,
        # left primer å’Œ right primer è¿›è¡Œ 3' ç«¯åå‘äº’è¡¥å½¢æˆå¼•ç‰©äºŒèšä½“
        'PRIMER_PAIR_MAX_COMPL_END': 3,
        'PRIMER_WT_PAIR_COMPL_END': 0.0,
        'PRIMER_PAIR_MAX_COMPL_END_TH': 35.0,
        'PRIMER_WT_PAIR_COMPL_END_TH': 302.4,
        # å‘å¤¹ç»“æ„ï¼Œç½šåˆ†ç³»æ•°è®¡ç®—ä¸º 9 / ( 1/(60-4-24) - 1/(60-4-0) ) = 672
        'PRIMER_MAX_HAIRPIN_TH': 24.0,
        'PRIMER_WT_HAIRPIN_TH': 672.0,
        # 3' ç«¯ç¢±åŸºçš„ç¨³å®šæ€§
        'PRIMER_MAX_END_STABILITY': 9,
        'PRIMER_WT_END_STABILITY': 1,
        ### ç¢±åŸºåºåˆ—è®¾å®šï¼š ###
        # æ¨¡æ¿åºåˆ—ä¸­åŒ…å«å°å†™å­—ç¬¦ä¸å½±å“å¼•ç‰©è®¾è®¡
        'PRIMER_LOWERCASE_MASKING': 0,
        # å¼•ç‰©åºåˆ—ä¸­ä¸èƒ½åŒ…å«å•æ ¸è‹·é…¸è¿ç»­é•¿åº¦è¶…è¿‡ 4 bp
        'PRIMER_MAX_POLY_X': 4,
        # å¼•ç‰©ä¸­å…è®¸çš„ N çš„æ•°ç›®
        'PRIMER_MAX_NS_ACCEPTED': 0,
        # æ¯ä¸ª N çš„ç½šåˆ†
        'PRIMER_WT_NUM_NS': 0.0,
        # å¼•ç‰©ä¸­ 3' ç«¯ 5bp ç¢±åŸºä¸­å…è®¸çš„æœ€å¤§ Gs æˆ– Cs çš„æ•°ç›®
        'PRIMER_MAX_END_GC': 4,
        # å¼•ç‰©ä¸­ 3' ç«¯ç¢±åŸºä¸­ä¸èƒ½å‡ºç°è¿ç»­çš„ Gs å’Œ Cs åºåˆ—
        'PRIMER_GC_CLAMP': 0,
        # æ˜¯å¦å…è®¸æœ‰è¯¸å¦‚ N A R Y ç­‰ç±»å‹çš„ç¢±åŸºã€‚å¿…é¡»åœ¨è®¾å®šPRIMER_MAX_NS_ACCEPTED ä¸ä¸º 0 åæ–¹æœ‰æ•ˆã€‚
        'PRIMER_LIBERAL_BASE': 1,
        # å¦‚æœè®¾ç½®ä¸º 1ï¼Œåˆ™ C èƒ½ä¸ S å®Œç¾åŒ¹é…ï¼Œä»»æ„ç¢±åŸºèƒ½å’Œ N å®Œç¾åŒ¹é…ã€‚
        'PRIMER_LIB_AMBIGUITY_CODES_CONSENSUS': 0,
        ### ç¢±åŸºè´¨é‡è®¾å®šï¼š ###
        # primer åºåˆ—å…è®¸æœ€å°çš„ç¢±åŸºè´¨é‡
        'PRIMER_MIN_QUALITY': 0,
        'PRIMER_MIN_END_QUALITY': 0,
        'PRIMER_QUALITY_RANGE_MIN': 0,
        'PRIMER_QUALITY_RANGE_MAX': 100,
        'PRIMER_WT_SEQ_QUAL': 0.0,
        'PRIMER_WT_END_QUAL': 0.0,
        ### PCR ååº”ä½“ç³»è®¾å®š: ###
        # å•ä»·ç›ç¦»å­æµ“åº¦(mM)
        'PRIMER_SALT_MONOVALENT': 50.0,
        # PRIMER_SALT_CORRECTIONS=1 means use the salt correction in SantaLucia et al 1998
        'PRIMER_SALT_CORRECTIONS': 1,
        # äºŒä»·é•ç¦»å­æµ“åº¦(mM)
        'PRIMER_SALT_DIVALENT': 1.5,
        # æ€»dNTPæµ“åº¦(mM)
        'PRIMER_DNTP_CONC': 0.6,
        # DNAäº§ç‰©æµ“åº¦(mM)
        'PRIMER_DNA_CONC': 50.0,
        ### PCR äº§ç‰©çš„è®¾å®šï¼š ###
        'PRIMER_PRODUCT_MIN_TM': -1000000.0,
        'PRIMER_PRODUCT_OPT_TM': 0.0,
        'PRIMER_PRODUCT_MAX_TM': 1000000.0,
        'PRIMER_PAIR_WT_PRODUCT_TM_LT': 0.0,
        'PRIMER_PAIR_WT_PRODUCT_TM_GT': 0.0,
        'PRIMER_PRODUCT_OPT_SIZE': 0,
        'PRIMER_PAIR_WT_PRODUCT_SIZE_LT': 0.0,
        'PRIMER_PAIR_WT_PRODUCT_SIZE_GT': 0.0,
        # ç½šåˆ†å› å­
        # left primerå’Œright primerçš„ç½šåˆ†ä¹‹å’Œã€‚æ­¤å’Œä¹˜ä»¥æ­¤ç³»æ•°ï¼Œå†åŠ å…¶å®ƒç½šåˆ†ä½œä¸ºæœ€ç»ˆç½šåˆ†ã€‚
        'PRIMER_PAIR_WT_PR_PENALTY': 1.0,
        # å°† internal oligo çš„ç½šåˆ†ä¹˜ä»¥æ­¤ç³»æ•°ï¼ŒåŠ å…¥åˆ°å¼•ç‰©çš„æœ€ç»ˆç½šåˆ†ä¸­ã€‚
        'PRIMER_PAIR_WT_IO_PENALTY': 0.0,
        ### internal oligo çš„è®¾å®šï¼š###
        # TM å€¼
        'PRIMER_INTERNAL_MIN_TM': 57.0,
        'PRIMER_INTERNAL_OPT_TM': 60.0,
        'PRIMER_INTERNAL_MAX_TM': 63.0,
        'PRIMER_INTERNAL_WT_TM_LT': 1.0,
        'PRIMER_INTERNAL_WT_TM_GT': 1.0,
        # é•¿åº¦
        'PRIMER_INTERNAL_MIN_SIZE': 18,
        'PRIMER_INTERNAL_OPT_SIZE': 20,
        'PRIMER_INTERNAL_MAX_SIZE': 27,
        'PRIMER_INTERNAL_WT_SIZE_LT': 1.0,
        'PRIMER_INTERNAL_WT_SIZE_GT': 1.0,
        # GC å«é‡
        'PRIMER_INTERNAL_MIN_GC': 20.0,
        'PRIMER_INTERNAL_MAX_GC': 80.0,
        'PRIMER_INTERNAL_OPT_GC_PERCENT': 50.0,
        'PRIMER_INTERNAL_WT_GC_PERCENT_LT': 0.0,
        'PRIMER_INTERNAL_WT_GC_PERCENT_GT': 0.0,
        # çƒ­åŠ›å­¦
        'PRIMER_INTERNAL_MAX_SELF_ANY': 12.00,
        'PRIMER_INTERNAL_WT_SELF_ANY': 0.0,
        'PRIMER_INTERNAL_MAX_SELF_END': 12.00,
        'PRIMER_INTERNAL_WT_SELF_END': 0.0,
        # ç¢±åŸºåºåˆ—
        'PRIMER_INTERNAL_MAX_POLY_X': 5,
        'PRIMER_INTERNAL_MAX_NS_ACCEPTED': 0,
        # ç¢±åŸºè´¨é‡
        'PRIMER_INTERNAL_MIN_QUALITY': 0,
        'PRIMER_INTERNAL_WT_END_QUAL': 0.0,
        'PRIMER_INTERNAL_WT_SEQ_QUAL': 0.0,
        # éå¼•ç‰©æ•°æ®åº“
        # PRIMER_INTERNAL_MAX_LIBRARY_MISHYB=12.00
        # PRIMER_INTERNAL_WT_LIBRARY_MISHYB=0.0
        # PCR ååº”ä½“ç³»è®¾å®š: ###
        # å•ä»·ç›ç¦»å­æµ“åº¦(mM)
        'PRIMER_INTERNAL_SALT_MONOVALENT': 50.0,
        'PRIMER_INTERNAL_SALT_DIVALENT': 1.5,
        # æ€»dNTPæµ“åº¦(mM)
        'PRIMER_INTERNAL_DNTP_CONC': 0.0,
        # DNAäº§ç‰©æµ“åº¦(mM)
        'PRIMER_INTERNAL_DNA_CONC': 50.0,
        ### PCR äº§ç‰©çš„è®¾å®šï¼š ###
        # 'PRIMER_PRODUCT_SIZE_RANGE': [140, 160],

    }
    return global_args


def check_options(parser):
    args = parser.parse_args()
    if not os.path.exists(args.input):
        parser.print_help()
        sys.exit(1)
    if not args.output:
        parser.print_help()
        sys.exit(1)
    return args


def get_options():
    parser = argparse.ArgumentParser(description="Primer3")
    parser.add_argument(
        "-i", "--input", help="fasta input file (seq name must uniq)", required=True)
    parser.add_argument(
        "-n", "--num", help="number of primers to be designed", default=1
    )
    parser.add_argument(
        "-min_size", "--min_size", help="minimal size of primers", default=18
    )
    parser.add_argument(
        "-max_size", "--max_size", help="maximal size of primers", default=22
    )
    parser.add_argument(
        "-min_tm", "--min_tm", help="minimal melting temperature of primers", default=55
    )
    parser.add_argument(
        "-max_tm", "--max_tm", help="maximal melting temperature of primers", default=65
    )
    parser.add_argument(
        "-min_gc", "--min_gc", help="minimal GC content of primers", default=40
    )
    parser.add_argument(
        "-max_gc", "--max_gc", help="maximal GC content of primers", default=70
    )
    parser.add_argument(
        "-t", "--thread", help="number of threads", default=1
    )
    parser.add_argument("-o", "--output", help="output file", required=True)
    return parser


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("\n\nCTRL-C detected! Exiting...")
        sys.exit(0)

```
é‚£ä¹ˆè¿è¡Œä¸€ä¸‹å§----
`python primer3_design.py -i SFSchr.fasta_second.fasta_vp1c_uniq.fasta -t 16 -o SFSchr.fasta_second.fasta_vp1c_uniq.out`

